zerver/lib/recipient_users.py

```diff
@@ -34,25 +35,36 @@ def get_recipient_from_user_profiles(
         if forwarder_user_profile.id not in recipient_profiles_map:
             raise ValidationError(_("User not authorized for this query"))
 
-    # If the direct message is just between the sender and
-    # another person, force it to be a personal internally
-    if len(recipient_profiles_map) == 2 and sender.id in recipient_profiles_map:
-        del recipient_profiles_map[sender.id]
+    # Make sure the sender is included in the group direct messages.
+    recipient_profiles_map[sender.id] = sender
+    user_ids = list(recipient_profiles_map)
+
+    # This is only for 1:1 or to-self direct messages.
+    # We want to use direct message group as an alternative if exists.
+    # Otherwise, we will use the legacty personal recipient.
+    if len(recipient_profiles_map) <= 2:
+        direct_message_group = get_direct_message_group(user_ids)
+        if direct_message_group:
+            # Use the existing direct message group as the preferred recipient.
+            return Recipient(
+                id=direct_message_group.recipient_id,
+                type=Recipient.DIRECT_MESSAGE_GROUP,
+                type_id=direct_message_group.id,
+            )
 
-    assert recipient_profiles_map
-    if len(recipient_profiles_map) == 1:
-        [user_profile] = recipient_profiles_map.values()
+        # if no direct message group recipient exists, we need to
+        # force the direct message to be a personal internally.
+        del recipient_profiles_map[sender.id]
+        if len(recipient_profiles_map) == 1:
+            [user_profile] = recipient_profiles_map.values()
+        else:
+            user_profile = sender
```

Comment:
I also realized that we need to add support to use the direct message group for self-messages. I also added [a test](https://github.com/zulip/zulip/pull/34307/files) for this scenario.

Reply:
Yeah, that is important.